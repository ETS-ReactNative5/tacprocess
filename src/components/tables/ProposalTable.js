import React from "react";
import propTypes from "prop-types";

const proposalsFilter = proposals => {
    const newProposals = []
    const longTermProposals = []
    proposals.map( p => {
      if (p.isNew){
        newProposals.push(p)
      }

      if (p.isLong) {
          longTermProposals.push(p)
      }
      return p
    })
    return {
      longTermProposals: longTermProposals,
      newProposals: newProposals
    };
  }

const getMatch = (a, b) => {
      const matches = [];

      for ( let i = 0; i < a.length; i++ ) {
          for ( let e = 0; e < b.length; e++ ) {
              if ( a[i] === b[e] ) matches.push( a[i] );
          }
      }
      return matches;
  }

class ProposalTable extends React.Component {
  render() {
    const { proposals} = this.props
    const newAndLong = proposalsFilter(proposals)
    return(
      <div>
        <table className="table left-table">
          <thead>
            <tr>
              <th />
              <th><h2>Total</h2></th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td><h3>New proposals plus older proposals</h3></td>
              <td><h3>{proposals.length}</h3></td>
            </tr>
            <tr>
              <td><h3>New proposals </h3></td>
              <td><h3>{newAndLong.newProposals.length}</h3></td>
            </tr>
            <tr>
              <td><h3>Older Proposals </h3></td>
              <td><h3>{proposals.length - newAndLong.newProposals.length}</h3></td>
            </tr>
            <tr>
              <td><h3>All long term proposals </h3></td>
              <td><h3>{newAndLong.longTermProposals.length}</h3></td>
            </tr>
            <tr>
              <td><h3>New long term proposals </h3></td>
              <td><h3>{getMatch(newAndLong.longTermProposals, newAndLong.newProposals).length}</h3></td>
            </tr>
          </tbody>
        </table>
      </div>
      );
    }
  }

  ProposalTable.propTypes = {
    proposals: propTypes.array.isRequired ,
  }

export default ProposalTable;
