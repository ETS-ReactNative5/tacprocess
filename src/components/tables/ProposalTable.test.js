import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import ProposalTable from "./ProposalTable";

const proposals = [
  {
    transparency: 'Clear',
    maxSeeing: 2,
    instruments: {
      scam: [ { dictatorMode: "NORMAL" } ],
    },
    timeRequests: [
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 150 }, { partnerCode: 'UW', time: 200 }, { partnerCode: 'IUCAA', time: 250 } ]
      },
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 400 }, { partnerCode: 'UW', time: 50 } ]
      }
    ]
  },
  {
    transparency: 'Thin cloud',
    maxSeeing: 2,
    instruments: {
      scam: [ { dictatorMode: "NORMAL" } ],
    },
    timeRequests: [
      {
        semester: '2017-1',
        distribution: [ { partnerCode: 'RSA', time: 150 }, { partnerCode: 'UW', time: 200 }, { partnerCode: 'IUCAA', time: 250 } ]
      },
      {
        semester: '2018-1',
        distribution: [ { partnerCode: 'RSA', time: 400 }, { partnerCode: 'UW', time: 50 } ]
      }
    ]
  }
];

// Checking if the ProposalTable Component renders correctly for different inputs
describe("ProposalTable Component", () => {

  //Enzyme method testing if it renders correctly with empty value of the target
  it("Render the ProposalTable Component having unpopulated props with no errors", () => {
    const rendered = mount(<ProposalTable proposals={[]} />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });

  it("Render the ProposalTable Component having populated props with no errors", () => {
    const rendered = mount(<ProposalTable proposals={proposals} />);
    expect(shallowToJson(rendered)).toMatchSnapshot();
  });

});
