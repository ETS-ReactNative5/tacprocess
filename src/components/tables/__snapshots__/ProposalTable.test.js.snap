// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalTable Component Render the ProposalTable Component having populated props with no errors 1`] = `
<ProposalTable
  proposals={
    Array [
      Object {
        "instruments": Object {
          "scam": Array [
            Object {
              "dictatorMode": "NORMAL",
            },
          ],
        },
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2018-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Clear",
      },
      Object {
        "instruments": Object {
          "scam": Array [
            Object {
              "dictatorMode": "NORMAL",
            },
          ],
        },
        "maxSeeing": 2,
        "timeRequests": Array [
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 150,
              },
              Object {
                "partnerCode": "UW",
                "time": 200,
              },
              Object {
                "partnerCode": "IUCAA",
                "time": 250,
              },
            ],
            "semester": "2017-1",
          },
          Object {
            "distribution": Array [
              Object {
                "partnerCode": "RSA",
                "time": 400,
              },
              Object {
                "partnerCode": "UW",
                "time": 50,
              },
            ],
            "semester": "2018-1",
          },
        ],
        "transparency": "Thin cloud",
      },
    ]
  }
>
  <div>
    <table
      className="table left-table"
    >
      <thead>
        <tr>
          <th />
          <th>
            <h2>
              Total
            </h2>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <h3>
              New proposals plus older proposals
            </h3>
          </td>
          <td>
            <h3>
              2
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              New proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Older Proposals 
            </h3>
          </td>
          <td>
            <h3>
              2
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              All long term proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              New long term proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</ProposalTable>
`;

exports[`ProposalTable Component Render the ProposalTable Component having unpopulated props with no errors 1`] = `
<ProposalTable
  proposals={Array []}
>
  <div>
    <table
      className="table left-table"
    >
      <thead>
        <tr>
          <th />
          <th>
            <h2>
              Total
            </h2>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <h3>
              New proposals plus older proposals
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              New proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              Older Proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              All long term proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
        <tr>
          <td>
            <h3>
              New long term proposals 
            </h3>
          </td>
          <td>
            <h3>
              0
            </h3>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</ProposalTable>
`;
